drivers: rtc: maxim,ds3231: convert to RTC API

gergovari opened the pull request:
This PR's goal is to convert the deprecated Counter-based API of the current driver to the newer RTC one.

I've not written actual driver code yet (that's why I made this a draft), as I wanted to see how Zephyr actually puts all the files together. (This is my first driver ever.)

Also, I already have a question: how are copyright notices handles? As you can see in the files I modified I put my name too as `Modified by ...` but I'm not sure what's the right move here.

Any mistakes I've made/improvement you see so far would be appreciated.

bjarki-andreasen commented:
Hi there :)

first of, I would not remove or alter the existing driver which uses the old API, both drivers can exist at the same time (we can remove the old one later) 

Regarding the copyright, if you substantially copy another file, add your copyright under the existing copyright in the new file :)

gregistech commented:
> Hi there :)
> 
> first of, I would not remove or alter the existing driver which uses the old API, both drivers can exist at the same time (we can remove the old one later) 
> 
> Regarding the copyright, if you substantially copy another file, add your copyright under the existing copyright in the new file :)

Like how the spi/mipi-dbi one works for displays in the new release? That sounds great, although I'm not sure yet how that's implemented. Probably the yaml files will point me in the right direction.

Thanks for clearing up copyright too :)

bjarki-andreasen commented on drivers/rtc/Kconfig.ds3231:
```diff
@@ -0,0 +1,16 @@
+	depends on DT_HAS_MAXIM_DS3231_ENABLED
+	depends on I2C
+	depends on RTC
+	select POLL
+	select NOTIFY
```
```
select POLL
select NOTIFY
```
Are not used by the driver, you can remove them :)

bjarki-andreasen commented on drivers/rtc/Kconfig.ds3231:
```diff
@@ -0,0 +1,16 @@
+	default y
+	depends on DT_HAS_MAXIM_DS3231_ENABLED
+	depends on I2C
+	depends on RTC
```
`depends on RTC` This one is not required as this file is only included if `CONFIG_RTC=y`

bjarki-andreasen reviewed (commented):
There is a test suite you can build which tests the RTC found at zephyr/tests/drivers/rtc/rtc_api which requires an overlay which uses the alias `rtc`. See https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#rtc-device-driver-test-suite, it can also test alarms and callbacks if those are implemented :)

Please squash all these commits and force push your branch so only one commit is present. Removing and re adding code in separate commits is not permitted, nor pretty :)

Using the RTC test, the sample should be unnecessary, especially when you can just build hello_world, and enable CONFIG_RTC, CONFIG_SHELL, and CONFIG_RTC_SHELL and play with it using the shell ;)

gregistech commented:
> There is a test suite you can build which tests the RTC found at zephyr/tests/drivers/rtc/rtc_api which requires an overlay which uses the alias `rtc`. See https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#rtc-device-driver-test-suite, it can also test alarms and callbacks if those are implemented :)
> Using the RTC test, the sample should be unnecessary, especially when you can just build hello_world, and enable CONFIG_RTC, CONFIG_SHELL, and CONFIG_RTC_SHELL and play with it using the shell ;)

I've been working on those too yesterday, plus some other features the DS3231 supports in addition to DS1307. I'll be sure to look into using a shell and tests!

> Please squash all these commits and force push your branch so only one commit is present. Removing and re adding code in separate commits is not permitted, nor pretty :)

Understandable, squashing should be easy.

gergovari committed:
drivers: rtc: maxim,ds3231: RTC driver

This is a squash of all the groundwork needed to
get a functioning driver for the DS3231 with the RTC API.

Signed-off-by: Gergo Vari <work@gergovari.com>

gregistech commented:
Will be on a vacation for a week, not sure if I get to write code in that period.
(Feel free to still write to me.)

For now I'd post the driver code as a gist for anyone to see the direction I'm going in.

It's probably not working atm.

https://gist.github.com/gregistech/9a6eecea8382c694f8eab61f8407aad2

bjarki-andreasen commented:
Gist looks good, enjoy your time off :)

gregistech commented:
@bjarki-andreasen how should I handle that:
- alarm_set_time weekfay and monthday can't be set at once
- and we either get an update callback every sec or we listen to the alarm interrupt?

These exclusive cases aren't really documented to my knowledge, how can I signal this to the user of the interface? (Obv. this is driver-specific)

bjarki-andreasen commented:
> - alarm_set_time weekday and monthday can't be set at once

Return -EINVAL if the user tries to set both weekday and monthday at once, the mask arg is used for this purpose.

> - and we either get an update callback every sec or we listen to the alarm interrupt?

Once the interrupt occurs, the status reg of the rtc should be read. The status will show what caused the interrupt, so it should be possible to support both.

gregistech commented:
> > - alarm_set_time weekday and monthday can't be set at once
> 
> Return -EINVAL if the user tries to set both weekday and monthday at once, the mask arg is used for this purpose.
> 
> > - and we either get an update callback every sec or we listen to the alarm interrupt?
> 
> Once the interrupt occurs, the status reg of the rtc should be read. The status will show what caused the interrupt, so it should be possible to support both.
> 
> 

Okay, the first is what I implemented already.

And yes on second thought I see how it will both work.

Thanks!

gregistech commented:
Actually, both interrupts can't be active.
```
When set to logic 1 with INTCN = 0 and VCC < VPF, this
bit enables the square wave. When BBSQW is logic 0,
the INT/SQW pin goes high impedance when VCC < VPF.
This bit is disabled (logic 0) when power is first applied.
```

BUT, I think this is what you had in mind: we can check every sec if an alarm flag got turned on.

kartben commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -0,0 +1,832 @@
+
+/* TODO: make sure we're in 24h mode */
+/* TODO: implement configurable settings */
+/* TODO: implement get_temp */
```
You need to look at devices that are implementing MFD (multifunction device). Essentially you would have a different *sensor* driver for what concerns the temperature, and the mfd compatible would allow you to lump the RTC and sensor under the same DT node.
See ex. https://github.com/zephyrproject-rtos/zephyr/blob/143b14bb4fb779437fdeb8369e42fbb81db4a5fd/boards/m5stack/m5stack_core2/m5stack_core2_procpu.dts#L119 for an MFD that acts as both a regulator and GPIO controller

gregistech replied:
Was about to ask this question now! Thank you for pointing me in the right direction.

gregvari commented:
*Just commited code for the sensor's API, not sure how to reopen the PR.*

kartben commented:
@gregvari just re-opened it for you

gregvari commented:
> @gregvari just re-opened it for you, although it doesn't look like you've actually pushed any change to the branch (yet?)

At first I accidentally pushed to main, and I thought I could fix it with rebasing and maybe moving the commit. From the look of the PR I shouldn't have as now I've other people's commits in my PR. Should I open a new PR or is there a way to fix the mess I created?

gregvari commented:
Now the read and decode api is done too, so if I'm correct these are the next things to be done:
- Ensure the I2C driver for the sensor is "RTIO ready"? (Not sure what this means, if it matters here or if that's the default behaviour.)
- Pass all CI checks
- Squash commits into one (correctly signed)

pdgendt commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -0,0 +1,868 @@
+}
+#endif /* defined(CONFIG_RTC_UPDATE) || defined(CONFIG_RTC_ALARM) */
+
+static const struct rtc_driver_api driver_api = {
```
Device driver APIs need to be wrapped now

```suggestion
static DEVICE_API(rtc, driver_api) = {
```

ttmut commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -0,0 +1,869 @@
+	tm->tm_nsec = 0;
+	tm->tm_isdst = -1;
+	tm->tm_yday = -1;
+	return 0;
```
This function never fails so it could be converted to void I guess.

ttmut commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -0,0 +1,869 @@
+		LOG_ERR("rtc_time_to_alarm_buf: Mask is invalid (%d)!\n", mask);
+		return -EINVAL;
+	}
+	if (id < 0 || id >= 2) {
```
```suggestion
	if (id < 0 || id >= ALARM_COUNT) {
```

bjarki-andreasen commented on drivers/mfd/mfd_ds3231.c:
```diff
@@ -0,0 +1,88 @@
+	return err;
+}
+
+bool mfd_ds3231_is_ready(const struct device *dev)
+{
+	const struct mfd_ds3231_conf *config = (struct mfd_ds3231_conf *)(dev->config);
+
+	return i2c_is_ready_dt(&(config->i2c_bus));
+}
+
+static int mfd_ds3231_init_i2c(const struct device *dev)
+{
+	struct mfd_ds3231_data *data = dev->data;
+
+	k_sem_init(&data->lock, 1, 1);
+	if (!mfd_ds3231_is_ready(dev)) {
+		LOG_ERR("I2C bus not ready.");
+		return -ENODEV;
+	}
+	return 0;
+}
+
+static int mfd_ds3231_init(const struct device *dev)
+{
+	return mfd_ds3231_init_i2c(dev);
+}
```
Since the sensor and RTC partial drivers reference the MFD struct device, `device_is_ready()` can be used directly instead if adding the new API `mfd_ds3231_is_ready()` since the i2c device readiness is already checked in `mfd_ds3231_init()`. If i2c device is not ready, `mfd_ds3231_init()` will fail and `device_is_ready()` on the mfd device will return false.

bjarki-andreasen commented on dts/bindings/rtc/maxim,ds3231-rtc.yaml:
```diff
@@ -0,0 +1,35 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+description: Maxim DS3231 I2C RTC/TCXO
```
Could you expand the descriptions to provide example devicetree layouts for the MFD device? Similar to https://github.com/zephyrproject-rtos/zephyr/blob/feaf21f80cb6ddb5dafa53056154707c18c6b2c2/dts/bindings/comparator/nordic%2Cnrf-comp.yaml#L10-L17 

I believe this may be essential for devs to understand how to use the device :)

bjarki-andreasen reviewed (changes requested):
Looks really good, I think we need to add examples to the dts bindings

bjarki-andreasen commented on drivers/rtc/rtc_ds3231.c:
```diff
@@ -0,0 +1,855 @@
+
+	const struct rtc_ds3231_conf *config = dev->config;
+	struct rtc_ds3231_data *data = dev->data;
+
```
The status of the mfd device should still be checked, using the device_is_ready() function
```
if (!device_is_ready(config->mfd)) {
        return -ENODEV;
}
```

bjarki-andreasen replied:
same for the sensor driver

gergovari commented on drivers/sensor/maxim/ds3231/ds3231.c:
```diff
@@ -0,0 +1,273 @@
+static int sensor_ds3231_init(const struct device *dev)
+{
+	const struct sensor_ds3231_conf *config = dev->config;
+	struct sensor_ds3231_data *data = dev->data;
```
True, I shouldn't rush this much yet. 😄
